sql-plugin/src/main/java/com/nvidia/spark/rapids/UnsafeRowToColumnarBatchIterator.java

```diff
@@ -0,0 +1,150 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids;
+
+import ai.rapids.cudf.ColumnVector;
+import ai.rapids.cudf.DType;
+import ai.rapids.cudf.HostColumnVector;
+import ai.rapids.cudf.HostColumnVectorCore;
+import ai.rapids.cudf.HostMemoryBuffer;
+import ai.rapids.cudf.NvtxColor;
+import ai.rapids.cudf.NvtxRange;
+import ai.rapids.cudf.Table;
+import org.apache.spark.TaskContext;
+import org.apache.spark.sql.catalyst.CudfUnsafeRow;
+import org.apache.spark.sql.catalyst.expressions.Attribute;
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow;
+import org.apache.spark.sql.execution.metric.SQLMetric;
+import org.apache.spark.sql.vectorized.ColumnarBatch;
+import scala.collection.Iterator;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+/**
+ * This class converts UnsafeRow instances to ColumnarBatches on the GPU through the magic of
+ * code generation. This just provides most of the framework a concrete implementation will
+ * be generated based off of the schema.
+ */
+public abstract class UnsafeRowToColumnarBatchIterator implements Iterator<ColumnarBatch> {
+  protected final Iterator<UnsafeRow> input;
+  protected UnsafeRow pending = null;
+  protected final int numRowsEstimate;
+  protected final long dataLength;
+  protected final DType[] rapidsTypes;
+  protected final SQLMetric totalTime;
+  protected final SQLMetric numInputRows;
+  protected final SQLMetric numOutputRows;
+  protected final SQLMetric numOutputBatches;
+
+  protected UnsafeRowToColumnarBatchIterator(
+      Iterator<UnsafeRow> input,
+      Attribute[] schema,
+      CoalesceGoal goal,
+      SQLMetric totalTime,
+      SQLMetric numInputRows,
+      SQLMetric numOutputRows,
+      SQLMetric numOutputBatches) {
+    this.input = input;
+    int sizePerRowEstimate = CudfUnsafeRow.getRowSizeEstimate(schema);
+    numRowsEstimate = (int)Math.max(1,
+        Math.min(Integer.MAX_VALUE - 1, goal.targetSizeBytes() / sizePerRowEstimate));
+    dataLength = ((long) sizePerRowEstimate) * numRowsEstimate;
+    rapidsTypes = new DType[schema.length];
+    for (int i = 0; i < schema.length; i++) {
+      rapidsTypes[i] = GpuColumnVector.getRapidsType(schema[i].dataType());
+    }
+    this.totalTime = totalTime;
+    this.numInputRows = numInputRows;
+    this.numOutputRows = numOutputRows;
+    this.numOutputBatches = numOutputBatches;
+  }
+
+  @Override
+  public boolean hasNext() {
+    return pending != null || input.hasNext();
+  }
+
+  @Override
+  public ColumnarBatch next() {
+    if (!hasNext()) {
+      throw new NoSuchElementException();
+    }
+    final int BYTES_PER_OFFSET = DType.INT32.getSizeInBytes();
+    int dataOffset;
+    int currentRow;
```

Comment:
Curious why these are forward-declared here, seems unnecessary.

Reply:
I think it has something to do with how I translated the code from scala to java and??? I'll fix it.